src/pact/v3/interaction/_base.py

```diff
@@ -266,11 +267,16 @@ def with_body(
                 If `None`, then the function intelligently determines whether
                 the body should be added to the request or the response.
         """
+        if not isinstance(body, str):
+            body_str = json.dumps(body, cls=MatcherEncoder)
```

Comment:
Nit: Avoid negative `isinstance` assertions.

If Python was strongly typed, this wouldn't matter; however, the type annotations are merely hint and at runtime, end users can in fact input whatever they want. So we need to handle cases of `with_body(MyClassWithouSerialization(...))` (which your `if` statement would catch).